README
## Pazpar2

Copyright (C) 2006-2017 Index Data.
See the file LICENSE for details.

Pazpar2 is a high-performance, user interface-independent, data
model-independent metasearching middleware featuring merging, relevance
ranking, record sorting, and faceted results.

Installation instructions and documentation is found in directory 'doc' and
[online](http://www.indexdata.com/pazpar2/doc) .

To get more information or assistance, send mail to yaz-help@indexdata.dk.
Even better, sign on to the
[YAZ mailing list](http://lists.indexdata.dk/cgi-bin/mailman/listinfo/yazlist)


NEWS
--- 1.14.1 2021/10/01

Update pazpar2-apache Debian package to work with conf-enabled
system.

nmake updates for ICU 68 + VS 2019

Use pkg-config to link with YAZ

--- 1.14.0 2019/01/31

Add parsing of MARC 264 to tmarc.xsl  PAZ-1054

Add MARC 245 $n, $p to "title-complete" in tmarc processing PAZ-1056

--- 1.13.1 2018/07/31

Update documentation for "version=2" parameter for show,
  termlist, and bytarget PAZ-1052

Improve handling of 852 locations in tmarc.xml stylesheet PAZ-1053

Fix missing close file handle for mmap #22

Fix uinit-var in call to ioctlsocket on Windows #21

Fix closing resources if http binding fails #20

--- 1.13.0 2017/11/14

Allow OR-matching or filter key-value pairs PAZ-1050

--- 1.12.13 2017/10/06

Correction for logic in tmarc.xsl regarding medium.
Fixes PAZ-1048. Also update tmarc3 test case.

Simplify code for dumping configuration PAZ-1047

--- 1.12.12 2017/05/08

Add term for "book (large print)" to medium termlist in tmarc.xsl

Package for Debian Stretch. Stop support for CentOS 5.

--- 1.12.11 2017/03/27

Fix segfault when using filters PAZ-1045

Use same ICU location as YAZ on Windows

Fix include sys/resource.h does not exist on Windows PAZ-1044

--- 1.12.10 2017/03/24

Update tmarc.xsl: Improve handling of holdings from MARC records

--- 1.12.9 2016/12/29

Command info also reports NMEM status

--- 1.12.8 2016/12/22

Log errors, when redis/memcached fails PAZ-1043

--- 1.12.7 2016/12/22

Record file log: use only fflush for non-NULL file ptr

Use YAZ' wrbuf_malloc_info rather than using mallinfo

Fix double free https://github.com/indexdata/pazpar2/issues/12

--- 1.12.6 2016/09/13

Continue fetching of records even in case of ingest failures PAZ-1042

Fixes and updates to documentation

--- 1.12.5 2015/10/19

Usage limit PAZ-1038
Guard against running out of sockets/file-descriptors. The limit may
be set with <socket max="number"/> in main configuration. If omitted,
the system default is used (getrlimit RLIMIT_NOFILE).

--- 1.12.4 2015/09/15

log msg cleanup PAZ-1037

Fix leak in record parsing
Occurred when empty data was produced and metadata did not have the
attribute "empty". The empty attribute was introduced in PAZ-858

Updates to pz:authentication documentation

Refer to full Apache product

Document info command. This command has been around since first release
in 2007. Command info includes versions, statistics, etc.
The undocumented server-status comand is gone and completely replaced by
command info which returns general information of the pazpar2 instance.

Minor updates to pazpar2_play manual

--- 1.12.3 2015/09/03

Make pazpar2_play part of Deb/RPM PAZ-1031

Allow predictable sessions (Option -R - for pazpar2) PAZ-1030

Fix pazpar2 recording broken PAZ-1029

At session destroy, log facet IDs in use

log level for http is http (not HTTP)

Fix 0 ptr reference for POST and no Content-Type PAZ-1028

Log levels http+iochan for tests. Wait longer time in valgrind mode.

--- 1.12.2 2015/08/31

Facility disable/logging of POSTed content PAZ-1027
Log level "post" controls whether POSTed content is logged. This
is enabled by default. To disable, use -v -post on the command line.

Explain the semantics of using option -d twice in pazpar2 man page.

Update ajaxdev.xml and make it part of the dist PAZ-1025

--- 1.12.1 2015/08/26

Fix pazpar2 crash when using pz:extendrecs PAZ-1024

--- 1.12.0 2015/08/17

Fix PAZPAR2_PPMUTEX_H incorrectly defined PAZ-1023

Fix windows: crash in xmlDocFormatDump PAZ-1021
Problem is the passed FILE*. We can not pass FILE pointers between
DLLs. So we just avoid dumping the XML config during start on Windows.

Use YAZ comstack features for IPv6 and non-blocking getaddrinfo. PAZ-927

Upgrade to debian DH version 9 PAZ-1020

Fix bytarget list not cleared for new search PAZ-1019

--- 1.11.3 2015/07/28

Do not re-search if facetlimit scan fails to find value PAZ-1013

Optimize single-target sortmap - keep record order PAZ-1917

position rec-icd with leading zeros PAZ-1018

Fix Bad HTTP request causes a crash PAZ-1012

--- 1.11.2 2015/07/21

Fix Can not make doc on mochi PAZ-1015

--- 1.11.1 2015/07/21

Fix PAZ-1009: Do not create local facets when using facetmap:split

--- 1.11.0 2015/07/06

Facet ID term map PAZ-1008
New setting pz:facetmap:split:name (refer to documentation)

--- 1.10.0 2015/06/16

Sort by retrieval PAZ-1006
Allows sorting by the order that records were retrieved from databases.

Response from bytarget command includes native query PAZ-1005

Use lx2.loc.gov:210/LCDB_MARC8 instead of z3950.loc.gov:7090/voyager PAZ-1003

--- 1.9.0 2015/05/23

New metadata facility "icurule" for normalizing metadata text PAZ-1002
This allows ICU normalization of text before in-cluster merging takes
place.

Heuristics to deal with Primo s/e/page PAZ-1001

Fix test_limit_limitmap fails PAZ-1000

--- 1.8.8 2015/04/20

Limitmap ccl: deal with quotes and backslash PAZ-999

Handle date range from a limit parameter PAZ-997
The CCL term is not quoted unless it includes blanks. This will ensure
that r=r is handled correctly. This changes behavior a bit. If term
should not be converted to @and, then instead of r=r one should use
r=o or r=3 instead.

Construct md-pages-number form page/endpage PAZ-998

--- 1.8.7 2015/03/27

solr-pz2.xsl: extend Solr transformation to emit license_name field PAZ-996

cf.xsl: forwards flat affiliation structure PAZ-992

parpar2: log YAZ version upon start PAZ-993

Update and fix documentation for preferred targets

Fix spelling of present_chunk in documentation

--- 1.8.6 2015/02/10

Increase facet limit PAZ-990
No longer a limit on facets.

Pazpar2-js no longer requires Apache2 PAZ-989
Package pazpar2-apache2 sets up everything WRT apache and only
package which requires it. pazpar2-apache depends on pazpar2-js.

--- 1.8.5 2014/12/17

Larger string buffer for forward address PAZ-987
Fixes buffer overflow that could occur if forward address is IPV6.


eventl.h
/*
 * Copyright (c) 1995-1999, Index Data
 * See the file LICENSE for details.
 * Sebastian Hammer, Adam Dickmeiss
 *
 * $Log: eventl.h,v $
 * Revision 1.3  2006-12-12 02:36:24  quinn
 * Implemented session timeout; ping command
 *
 * Revision 1.2  2006/11/18 05:00:38  quinn
 * Added record retrieval, etc.
 *
 * Revision 1.1.1.1  2006/11/14 20:44:38  quinn
 * PazPar2
 *
 * Revision 1.1.1.1  2000/02/23 14:40:18  heikki
 * Original import to cvs
 *
 * Revision 1.11  1999/04/20 09:56:48  adam
 * Added 'name' paramter to encoder/decoder routines (typedef Odr_fun).
 * Modified all encoders/decoders to reflect this change.
 *
 * Revision 1.10  1998/01/29 13:30:23  adam
 * Better event handle system for NT/Unix.
 *
 * Revision 1.9  1997/09/01 09:31:48  adam
 * Removed definition statserv_remove from statserv.h to eventl.h.
 *
 * Revision 1.8  1995/06/19 12:39:09  quinn
 * Fixed bug in timeout code. Added BER dumper.
 *
 * Revision 1.7  1995/06/16  10:31:34  quinn
 * Added session timeout.
 *
 * Revision 1.6  1995/05/16  08:51:02  quinn
 * License, documentation, and memory fixes
 *
 * Revision 1.5  1995/05/15  11:56:37  quinn
 * Asynchronous facilities. Restructuring of seshigh code.
 *
 * Revision 1.4  1995/03/27  08:34:23  quinn
 * Added dynamic server functionality.
 * Released bindings to session.c (is now redundant)
 *
 * Revision 1.3  1995/03/15  08:37:42  quinn
 * Now we're pretty much set for nonblocking I/O.
 *
 * Revision 1.2  1995/03/14  10:28:00  quinn
 * More work on demo server.
 *
 * Revision 1.1  1995/03/10  18:22:45  quinn
 * The rudiments of an asynchronous server.
 *
 */

#ifndef EVENTL_H
#define EVENTL_H

#include <time.h>

struct iochan;

typedef void (*IOC_CALLBACK)(struct iochan *i, int event);

typedef struct iochan
{
    int fd;
    int flags;
#define EVENT_INPUT     0x01
#define EVENT_OUTPUT    0x02
#define EVENT_EXCEPT    0x04
#define EVENT_TIMEOUT   0x08
#define EVENT_WORK      0x10
    int force_event;
    IOC_CALLBACK fun;
    void *data;
    int destroyed;
    time_t last_event;
    time_t max_idle;
    
    struct iochan *next;
} *IOCHAN;

#define iochan_destroy(i) (void)((i)->destroyed = 1)
#define iochan_getfd(i) ((i)->fd)
#define iochan_setfd(i, f) ((i)->fd = (f))
#define iochan_getdata(i) ((i)->data)
#define iochan_setdata(i, d) ((i)->data = d)
#define iochan_getflags(i) ((i)->flags)
#define iochan_setflags(i, d) ((i)->flags = d)
#define iochan_setflag(i, d) ((i)->flags |= d)
#define iochan_clearflag(i, d) ((i)->flags &= ~(d))
#define iochan_getflag(i, d) ((i)->flags & d ? 1 : 0)
#define iochan_getfun(i) ((i)->fun)
#define iochan_setfun(i, d) ((i)->fun = d)
#define iochan_setevent(i, e) ((i)->force_event = (e))
#define iochan_getnext(i) ((i)->next)
#define iochan_settimeout(i, t) ((i)->max_idle = (t), (i)->last_event = time(0))
#define iochan_activity(i) ((i)->last_event = time(0))

IOCHAN iochan_create(int fd, IOC_CALLBACK cb, int flags);
int event_loop(IOCHAN *iochans);

#endif
